hail/src/main/scala/is/hail/expr/ir/ExtractIntervalFilters.scala

```diff
@@ -1,360 +1,909 @@
+          BoolValue.fromComparisonKeyPrefix(l.asInstanceOf[Row], op).restrict(keySet)
+        case _ => BoolValue.top(keySet)
+      }
+    }
+
+    private def opIsSupported(op: ComparisonOp[_]): Boolean = op match {
+      case _: EQ | _: NEQ | _: LTEQ | _: LT | _: GTEQ | _: GT | _: EQWithNA | _: NEQWithNA => true
       case _ => false
     }
   }
 
-  def literalSizeOkay(lit: Literal): Boolean = lit.value.asInstanceOf[Iterable[_]].size <= MAX_LITERAL_SIZE
+  import Lattice.{ Value => AbstractValue, ConstantValue, KeyField, StructValue, BoolValue, Contig, Position }
 
-  def wrapInRow(intervals: Array[Interval]): Array[Interval] = {
-    intervals.map { interval =>
-      Interval(IntervalEndpoint(Row(interval.left.point), interval.left.sign),
-        IntervalEndpoint(Row(interval.right.point), interval.right.sign))
-    }
+  case class AbstractEnv(keySet: KeySet, env: Env[AbstractValue]) {
+    def lookupOption(name: String): Option[AbstractValue] =
+      env.lookupOption(name)
+
+    def bind(bindings: (String, AbstractValue)*): AbstractEnv =
+      copy(env = env.bind(bindings: _*))
+
+    def restrict(k: KeySet): AbstractEnv =
+      copy(keySet = KeySetLattice.meet(keySet, k))
   }
 
-  def minimumValueByType(t: Type, es: ExtractionState): Any = {
-    t match {
-      case TInt32 => Int.MinValue
-      case TInt64 => Long.MinValue
-      case TFloat32 => Float.NegativeInfinity
-      case TFloat64 => Double.PositiveInfinity
-      case TBoolean => false
-      case t: TLocus =>
-        val rg = es.getReferenceGenome(t.rg)
-        Locus(rg.contigs.head, 1)
-      case tbs: TBaseStruct => Row.fromSeq(tbs.types.map(minimumValueByType(_, es)))
+  object AbstractEnvLattice extends JoinLattice {
+    type Value = AbstractEnv
+    val top: Value = AbstractEnv(KeySetLattice.top, Env.empty)
+
+    def combine(l: Value, r: Value): Value = {
+      val keys = l.env.m.keySet.intersect(r.env.m.keySet)
+      val env = Env.fromSeq(keys.map { k =>
+        k -> Lattice.combine(l.env(k), r.env(k))
+      })
+      AbstractEnv(KeySetLattice.combine(l.keySet, r.keySet), env)
     }
   }
 
-  def maximumValueByType(t: Type, es: ExtractionState): Any = {
-    t match {
-      case TInt32 => Int.MaxValue
-      case TInt64 => Long.MaxValue
-      case TFloat32 => Float.PositiveInfinity
-      case TFloat64 => Double.PositiveInfinity
-      case TBoolean => false
-      case t: TLocus =>
-        val rg = es.getReferenceGenome(t.rg)
-        val contig = rg.contigs.last
-        Locus(contig, rg.contigLength(contig) - 1)
-      case tbs: TBaseStruct => Row.fromSeq(tbs.types.map(maximumValueByType(_, es)))
+  def firstKeyOrd: ExtendedOrdering = keyType.types.head.ordering(ctx.stateManager)
+  def keyOrd: ExtendedOrdering = PartitionBoundOrdering(ctx, keyType)
+  val iord: IntervalEndpointOrdering = keyOrd.intervalEndpointOrdering
+
+  private def intervalsFromLiteral(lit: Any, ordering: Ordering[Any], wrapped: Boolean): IndexedSeq[Interval] =
```

Comment:
I think it would be better to use the type alias `KeySet` here instead of `IndexedSeq[Interval]` as this helps make changes in the future. 

Say for example, you defined the values of the keysetlattice like this:
```scala
sealed trait KeySet extends Product with Serializable
case object Top extends KeySet
case class Keys(intervals: IndexedSeq[Interval]) extends KeySet
case object Bottom extends KeySet
```

Such an implementation is a direct swap-in when you use the type alias.

Reply:
Good catch. At points during refactoring things, these lived outside the scope of `KeySet`.